src/BotPlutusInterface/Balance.hs

```diff
@@ -65,71 +69,127 @@ import Plutus.V1.Ledger.Api (
+  (Member (PABEffect w) effs) =>
+  PABConfig ->
+  PubKeyHash ->
+  UnbalancedTx ->
+  Eff effs (Either Text Tx)
+balanceTxIO pabConf ownPkh unbalancedTx@(UnbalancedTx tx' _ _ _)
+  | validCollateralTx tx' = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxCollateral
+  | txUsesScripts tx' = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxWithScripts
+  | otherwise = balanceTxIO' @w pabConf ownPkh unbalancedTx BalanceTxWithoutScripts
+  where
+    validCollateralTx :: Tx -> Bool
+    validCollateralTx tx
+      | [out] <- txOutputs tx
+        , Map.size (txData tx) == 0
+        , txMint tx == mempty
+        , txOutValue out == collateralValue pabConf
+        , txMintScripts tx == mempty
+        , txCollateral tx == mempty
+        , not (txUsesScripts tx) =
+        True
+      | otherwise =
+        False
+
 {- | Collect necessary tx inputs and collaterals, add minimum lovelace values and balance non ada
- assets
+     assets
 -}
-balanceTxIO ::
+balanceTxIO' ::
   forall (w :: Type) (effs :: [Type -> Type]).
-  Member (PABEffect w) effs =>
+  (Member (PABEffect w) effs) =>
   PABConfig ->
   PubKeyHash ->
   UnbalancedTx ->
+  BalanceTx ->
   Eff effs (Either Text Tx)
-balanceTxIO pabConf ownPkh unbalancedTx =
+balanceTxIO' pabConf ownPkh unbalancedTx balanceTxType =
   runEitherT $
     do
-      utxos <- newEitherT $ CardanoCLI.utxosAt @w pabConf changeAddr
+      (utxos, mcollateral) <- newEitherT $ utxosAndCollateralAtAddress @w balanceTxType pabConf changeAddr
       privKeys <- newEitherT $ Files.readPrivateKeys @w pabConf
+
       let utxoIndex = fmap Tx.toTxOut utxos <> unBalancedTxUtxoIndex unbalancedTx
           requiredSigs = map Ledger.unPaymentPubKeyHash $ Map.keys (unBalancedTxRequiredSignatories unbalancedTx)
 
+      lift $ printBpiLog @w Debug $ viaShow utxoIndex
+
+      -- We need this folder on the CLI machine, which may not be the local machine
+      lift $ createDirectoryIfMissingCLI @w False (Text.unpack "pcTxFileDir")
+
       tx <-
         newEitherT $
           addValidRange @w
             (unBalancedTxValidityTimeRange unbalancedTx)
             (unBalancedTxTx unbalancedTx)
 
-      lift $ printBpiLog @w Debug $ viaShow utxoIndex
-
-      -- We need this folder on the CLI machine, which may not be the local machine
-      lift $ createDirectoryIfMissingCLI @w False (Text.unpack pabConf.pcTxFileDir)
-
-      -- Adds required collaterals, only needs to happen once
+      -- Adds required collaterals in the `Tx`, if the `Tx` is of type `BalanceTxWithScripts`.
       -- Also adds signatures for fee calculation
-      preBalancedTx <- hoistEither $ addTxCollaterals utxoIndex tx >>= addSignatories ownPkh privKeys requiredSigs
+      preBalancedTx <-
+        if balanceTxType == BalanceTxWithScripts
+          then
+            hoistEither $
+              addSignatories ownPkh privKeys requiredSigs $
+                addTxCollaterals (fromJust mcollateral) tx
```

Comment:
I'd avoid `fromJust` as well, since the error can't be caught. Why not convert the `Maybe` (i.e. `mcollateral`) to an `Either` and use `hoistEither` above here (after line 129)?

Reply:
fixed